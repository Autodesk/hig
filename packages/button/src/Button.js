import React, { Component } from "react";
import PropTypes from "prop-types";
import { ControlBehavior } from "@hig/behaviors";

import {
  availableTargets,
  availableTypes,
  availableWidths,
  types,
  widths
} from "./constants";
import ButtonPresenter from "./presenters/ButtonPresenter";

export default class Button extends Component {
  static propTypes = {
    /**
     * Prevents user interaction with the button
     */
    disabled: PropTypes.bool,
    /**
     * A @hig/icon element
     */
    icon: PropTypes.node,
    /**
     * Sets the link of a button
     */
    link: PropTypes.string,
    /**
     * Triggers blur when focus is moved away from icon
     */
    onBlur: PropTypes.func,
    /**
     * Triggers when you click the button
     */
    onClick: PropTypes.func,
    /**
     * Triggers when focus is moved to button
     */
    onFocus: PropTypes.func,
    /**
     * Triggers when you hover over the button
     */
    onHover: PropTypes.func,
    /**
     * Triggers when the user's mouse is pressed over the button
     */
    onMouseDown: PropTypes.func,
    /**
     * Triggers when the user's mouse is over the button
     */
    onMouseEnter: PropTypes.func,
    /**
     * Triggers when the user's mouse is no longer over the button
     */
    onMouseLeave: PropTypes.func,
    /**
     * Triggers when the user's mouse is no longer pressed over the button
     */
    onMouseUp: PropTypes.func,
    /**
     * Adds custom/overriding styles
     */
    stylesheet: PropTypes.func,
    /**
     * Specifies where to display the linked URL
     */
    target: PropTypes.oneOf(availableTargets),
    /**
     * Sets the title of a button
     */
    title: PropTypes.string.isRequired,
    /**
     * Specifies type of button
     */
    type: PropTypes.oneOf(availableTypes),
    /**
     * Specifies width of button
     */
    width: PropTypes.oneOf(availableWidths)
  };

  static defaultProps = {
    disabled: false,
    type: types.SOLID,
    width: widths.SHRINK
  };

  render() {
    const {
      disabled,
      icon,
      link,
      onBlur,
      onClick,
      onFocus,
      onHover,
      onMouseDown,
      onMouseEnter,
      onMouseLeave,
      onMouseUp,
      stylesheet,
      target,
      title,
      type,
      width,
      ...otherProps
    } = this.props;

    return (
      <ControlBehavior
        onBlur={onBlur}
        onFocus={onFocus}
        onMouseDown={onMouseDown}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onMouseUp={onMouseUp}
      >
        {({
          hasFocus,
          hasHover,
          isPressed,
          onBlur: handleBlur,
          onFocus: handleFocus,
          onMouseDown: handleMouseDown,
          onMouseEnter: handleMouseEnter,
          onMouseLeave: handleMouseLeave,
          onMouseUp: handleMouseUp
        }) => (
          <ButtonPresenter
            disabled={disabled}
            hasFocus={hasFocus}
            hasHover={hasHover}
            isPressed={isPressed}
            icon={icon}
            link={link}
            onBlur={handleBlur}
            onClick={onClick}
            onFocus={handleFocus}
            onHover={onHover}
            onMouseDown={handleMouseDown}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseLeave}
            onMouseUp={handleMouseUp}
            stylesheet={stylesheet}
            target={target}
            title={title}
            type={type}
            width={width}
            {...otherProps}
          />
        )}
      </ControlBehavior>
    );
  }
}
