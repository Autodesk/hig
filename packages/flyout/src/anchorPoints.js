export const anchorPoints = Object.freeze({
  TOP_LEFT: "top-left",
  TOP_CENTER: "top-center",
  TOP_RIGHT: "top-right",
  RIGHT_TOP: "right-top",
  RIGHT_CENTER: "right-center",
  RIGHT_BOTTOM: "right-bottom",
  BOTTOM_LEFT: "bottom-left",
  BOTTOM_CENTER: "bottom-center",
  BOTTOM_RIGHT: "bottom-right",
  LEFT_TOP: "left-top",
  LEFT_CENTER: "left-center",
  LEFT_BOTTOM: "left-bottom",
});

export const AVAILABLE_ANCHOR_POINTS = Object.freeze(
  Object.values(anchorPoints)
);
