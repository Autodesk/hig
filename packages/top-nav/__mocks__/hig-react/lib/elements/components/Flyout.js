import React from "react";

function FlyoutMock({ content, children }) {
  return (
    <div data-mock="flyout">
      <div data-mock="flyout-content">{content}</div>
      <div data-mock="flyout-children">{children}</div>
    </div>
  );
}

module.exports = jest.fn(FlyoutMock);
