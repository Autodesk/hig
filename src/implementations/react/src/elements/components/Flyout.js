import * as PropTypes from "prop-types";
import React, { Component } from "react";
import * as HIG from 'hig-vanilla';
import FlyoutAdapter from "../../adapters/NewFlyoutAdapter";

class Flyout extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    };
  }

  static propTypes = {
    anchorPoint: PropTypes.oneOf(HIG.Flyout.AvailableAnchorPoints),
    children: PropTypes.node,
    content: PropTypes.node
  }

  static __docgenInfo = {
    props: {
      anchorPoint: {
        description: "where the flyout will be anchored relative to target"
      },
      children: { description: "target component to open the flyout" },
      content: { description: "content for the flyout" },
    }
  };

  closeFlyout = () => {
    this.setState({ open: false });
  };

  openFlyout = event => {
    console.log('CLICKING THE THING');
    event.preventDefault();
    this.setState({ open: true });
  };

  render() {
    const target = this.props.children
      ? React.cloneElement(this.props.children, { onClick: this.openFlyout })
      : null
    return (
      <FlyoutAdapter
        anchorPoint={this.props.anchorPoint}
        open={this.state.open}
        onClickOutside={this.closeFlyout}
        content={this.props.content}
      >
        {target}
      </FlyoutAdapter>
    );
  }
}

export default Flyout;
