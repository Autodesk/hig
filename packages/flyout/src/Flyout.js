import React, { Component } from "react";
import PropTypes from "prop-types";
import Transition from "react-transition-group/Transition";
import { polyfill } from "react-lifecycles-compat";

import { anchorPoints, availableAnchorPoints } from "./anchorPoints";
import FlyoutPresenter from "./FlyoutPresenter";
import getFlyoutPosition from "./getFlyoutPosition";

const TRANSITION_DURATION = 300;

class Flyout extends Component {
  static propTypes = {
    /** Where the flyout will be anchored relative to target */
    anchorPoint: PropTypes.oneOf(availableAnchorPoints),
    /** Target component to open the flyout. Can be either a node or a render function */
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    /** Content for the flyout */
    content: PropTypes.node,
    /** Max height of the flyout content, in pixels */
    maxHeight: PropTypes.number,
    /** Function called when the flyout is open, and a click event occurs outside the flyout */
    onClickOutside: PropTypes.func,
    /** Function called when the flyout panel is scrolled */
    onScroll: PropTypes.func,
    /** When provided, it overrides the flyout's open state */
    open: PropTypes.bool
  };

  static defaultProps = {
    anchorPoint: anchorPoints.RIGHT_TOP
  };

  /**
   * @param {FlyoutProps} nextProps
   * @param {FlyoutState} prevState
   * @returns {FlyoutState | null}
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    const { open } = nextProps;
    const { isVisible } = prevState;

    if (open === isVisible) return null;

    return { isVisible: open };
  }

  state = {
    isVisible: false
  };

  componentDidMount() {
    window.document.body.addEventListener("click", this.handleBodyClick);
  }

  componentWillUnmount() {
    window.document.body.removeEventListener("click", this.handleBodyClick);
  }

  getPresenterPositionProps() {
    const { action, panel } = this;

    if (!action || !panel) return { top: 0, left: 0 };

    const actionRect = action.getBoundingClientRect();
    const panelRect = panel.getBoundingClientRect();
    const viewportRect = window.document.documentElement.getBoundingClientRect();
    const { anchorPoint } = this.props;

    return getFlyoutPosition({
      anchorPoint,
      actionRect,
      panelRect,
      viewportRect
    });
  }

  /** @type {HTMLElement} */
  action;

  /** @type {HTMLDivElement} */
  container;

  /** @type {HTMLDivElement} */
  panel;

  /** @type {HTMLDivElement} */
  wrapper;

  handleChildClick = () => {
    this.toggleVisibility();
  };

  /**
   * @param {MouseEvent} event
   */
  handleBodyClick = event => {
    const { wrapper } = this;
    const { isVisible } = this.state;
    const { onClickOutside } = this.props;
    const flyoutClicked =
      event.target === wrapper || wrapper.contains(event.target);

    if (flyoutClicked || !isVisible) return;
    if (onClickOutside) onClickOutside(event);

    this.toggleVisibility();
  };

  /**
   * @param {HTMLElement} action
   */
  refAction = action => {
    this.action = action;
  };

  /**
   * @param {HTMLDivElement} container
   */
  refContainer = container => {
    this.container = container;
  };

  /**
   * @param {HTMLDivElement} panel
   */
  refPanel = panel => {
    this.panel = panel;
  };

  /**
   * @param {HTMLDivElement} wrapper
   */
  refWrapper = wrapper => {
    this.wrapper = wrapper;
  };

  toggleVisibility() {
    this.setState({
      isVisible: !this.state.isVisible
    });
  }

  createPresenterProps(transitionStatus) {
    const { content, maxHeight, onScroll } = this.props;
    const { refContainer, refPanel, refAction, refWrapper } = this;
    const {
      anchorPoint,
      topOffset,
      leftOffset
    } = this.getPresenterPositionProps();

    return {
      anchorPoint,
      topOffset,
      leftOffset,
      content,
      maxHeight,
      refAction,
      refContainer,
      refPanel,
      refWrapper,
      transitionStatus,
      onScroll
    };
  }

  renderChildren() {
    const { children } = this.props;
    const { handleChildClick } = this;

    if (typeof children === "function") {
      return children({ handleClick: handleChildClick });
    }

    if (React.Children.count(children) === 1) {
      return React.cloneElement(children, { onClick: handleChildClick });
    }

    return children;
  }

  renderPresenter = transitionStatus => (
    <FlyoutPresenter {...this.createPresenterProps(transitionStatus)}>
      {this.renderChildren()}
    </FlyoutPresenter>
  );

  render() {
    return (
      <Transition in={this.state.isVisible} timeout={TRANSITION_DURATION}>
        {this.renderPresenter}
      </Transition>
    );
  }
}

export default polyfill(Flyout);
