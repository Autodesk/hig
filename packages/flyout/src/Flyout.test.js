import React from "react";
import renderer from "react-test-renderer";
import Button from "@hig/button";

import { anchorPoints } from "./anchorPoints";
import Flyout from "./Flyout";

describe("flyout/Flyout", () => {
  [
    {
      description: "renders without props",
      props: {}
    },
    {
      description: "renders with all props",
      props: {
        anchorPoint: anchorPoints.TOP_CENTER,
        children: <Button title="Hello" />,
        content: "World",
        maxHeight: 150,
        onClickOutside: function onClickOutside() {},
        onScroll: function onScroll() {},
        open: true
      }
    },
    {
      description: "renders children from the given render function",
      props: {
        children: ({ handleClick }) => (
          <Button onClick={handleClick} title="Render prop" />
        )
      }
    }
  ].forEach(({ description, props: { children, ...otherProps } }) => {
    it(description, () => {
      const presenter = <Flyout {...otherProps}>{children}</Flyout>;
      const tree = renderer.create(presenter).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
