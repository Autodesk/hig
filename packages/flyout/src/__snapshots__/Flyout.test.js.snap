// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flyout/Flyout renders children from the given render function 1`] = `
<div
  className="hig__flyout-v1 hig__flyout-v1--exited hig__flyout-v1--top-right"
>
  <div
    className="hig__flyout-v1__action"
  >
    <button
      className="hig--light-theme hig__button-v1 hig__button-v1--size-standard hig__button-v1--type-primary hig__button-v1--width-shrink"
      onBlur={undefined}
      onClick={[Function]}
      onFocus={undefined}
      onMouseOver={undefined}
    >
      <span
        className="hig__button__title"
      >
        Render prop
      </span>
    </button>
  </div>
  <div
    className="hig__flyout-v1__container"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  >
    <div
      className="hig__flyout-v1__chevron"
    />
    <div
      className="hig__flyout-v1__panel"
      onScroll={undefined}
      style={
        Object {
          "maxHeight": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`flyout/Flyout renders with all props 1`] = `
<div
  className="hig__flyout-v1 hig__flyout-v1--top-right"
>
  <div
    className="hig__flyout-v1__action"
  >
    <button
      className="hig--light-theme hig__button-v1 hig__button-v1--size-standard hig__button-v1--type-primary hig__button-v1--width-shrink"
      onBlur={undefined}
      onClick={[Function]}
      onFocus={undefined}
      onMouseOver={undefined}
    >
      <span
        className="hig__button__title"
      >
        Hello
      </span>
    </button>
  </div>
  <div
    className="hig__flyout-v1__container"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  >
    <div
      className="hig__flyout-v1__chevron"
    />
    <div
      className="hig__flyout-v1__panel"
      onScroll={[Function]}
      style={
        Object {
          "maxHeight": 150,
        }
      }
    >
      World
    </div>
  </div>
</div>
`;

exports[`flyout/Flyout renders without props 1`] = `
<div
  className="hig__flyout-v1 hig__flyout-v1--exited hig__flyout-v1--top-right"
>
  <div
    className="hig__flyout-v1__action"
  />
  <div
    className="hig__flyout-v1__container"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  >
    <div
      className="hig__flyout-v1__chevron"
    />
    <div
      className="hig__flyout-v1__panel"
      onScroll={undefined}
      style={
        Object {
          "maxHeight": undefined,
        }
      }
    />
  </div>
</div>
`;
